rust/lance-index/src/scalar/bloomfilter.rs

```diff
@@ -0,0 +1,1769 @@
+            if self.blocks.is_empty() && self.cur_block_offset == 0 {
+                let first_row_addr = row_addrs_array.value(0);
+                self.cur_fragment_id = first_row_addr >> 32;
+            }
+
+            while remaining > 0 {
+                // Find the next fragment boundary in this batch
+                let next_fragment_index = (array_offset..row_addrs_array.len()).find(|&i| {
+                    let row_addr = row_addrs_array.value(i);
+                    let fragment_id = row_addr >> 32;
+                    fragment_id == self.cur_fragment_id + 1
+                });
+                let empty_rows_left_in_cur_zone: usize =
+                    (self.params.number_of_items - self.cur_block_offset as u64) as usize;
+
+                // Check if there is enough data from the current fragment to fill the current zone
+                let desired = if let Some(idx) = next_fragment_index {
+                    self.cur_fragment_id = row_addrs_array.value(idx) >> 32;
+                    // Take the minimum between distance to boundary and space left in zone
+                    // to ensure we don't exceed the zone size limit
+                    std::cmp::min(idx - array_offset, empty_rows_left_in_cur_zone)
+                } else {
+                    empty_rows_left_in_cur_zone
+                };
+
+                if desired > remaining {
+                    // Not enough data to fill a map, just increment counts
+                    self.update_stats(&data_array.slice(array_offset, remaining))?;
+                    self.cur_block_offset += remaining;
+                    break;
+                } else if desired > 0 {
+                    // There is enough data, create a new block
+                    self.update_stats(&data_array.slice(array_offset, desired))?;
+                    self.cur_block_offset += desired;
+                    self.new_block(row_addrs_array.value(array_offset) >> 32)?;
+                } else if desired == 0 {
+                    // The new batch starts with a new fragment. Flush the current zone if it's not empty
+                    if self.cur_block_offset > 0 {
+                        self.new_block(self.cur_fragment_id - 1)?;
+                    }
+                    // Let the loop run again
+                    // to find the next fragment boundary
+                    continue;
+                }
+                array_offset += desired;
+                remaining = remaining.saturating_sub(desired);
+            }
+        }
+        // Create the final block
+        if self.cur_block_offset > 0 {
+            self.new_block(self.cur_fragment_id)?;
+        }
+
+        Ok(())
+    }
+
+    fn bloomfilter_stats_as_batch(&self) -> Result<RecordBatch> {
+        // Create placeholder array for the indexed data type with same length as other columns
+        let data_type_placeholder = if self.blocks.is_empty() {
```

Comment:
why do we need to record data type? feels unnecssary

Reply:
Removed